fastavro/_write.pyx

```diff
@@ -295,6 +295,13 @@ cdef write_union(bytearray fo, datum, schema, dict named_schemas, fname):
                     continue
 
             if _validate(datum, candidate, named_schemas, raise_errors=False):
+                if isinstance(candidate, dict):
```

Comment:
I saw the part about the check being done here for the general case, but I think we can safely keep this logic in the specific record case. Therefore, could you move this down under the `if record_type == "record":` block? We can also get rid of the `if isinstance(candidate, dict):` check at that point.